### PR TITLE
NETOBSERV-1625: Add ebpf altering for flows drop

### PR DESCRIPTION
**Pull request by msherif1234**
## Description

Add alert for ebpf flows drop to detect when ebpf hash table is full

## Dependencies

n/a

## Checklist

If you are not familiar with our processes or don't know what to answer in the list below, let us know in a comment: the maintainers will take care of that.

* [X] Is this PR backed with a JIRA ticket? If so, make sure it is written as a title prefix _(in general, PRs affecting the NetObserv/Network Observability product should be backed with a JIRA ticket - especially if they bring user facing changes)._
* [X] Does this PR require product documentation?
  * [ ] If so, make sure the JIRA epic is labelled with "documentation" and provides a description relevant for doc writers, such as use cases or scenarios. Any required step to activate or configure the feature should be documented there, such as new CRD knobs.
* [ ] Does this PR require a product release notes entry?
  * [ ] If so, fill in "Release Note Text" in the JIRA.
* [ ] Is there anything else the QE team should know before testing? E.g: configuration changes, environment setup, etc.
  * [ ] If so, make sure it is described in the JIRA ticket.
* QE requirements (check 1 from the list):
  * [X] Standard QE validation, with pre-merge tests unless stated otherwise.
  * [ ] Regression tests only (e.g. refactoring with no user-facing change).
  * [ ] No QE (e.g. trivial change with high reviewer's confidence, or per agreement with the QE team).

### CODE REVIEW

**Comment by msherif1234** [comment]
```yaml
$ oc get prometheusrules -n netobserv-privileged ebpf-agent-prom-alert -o yaml
apiVersion: monitoring.coreos.com/v1
kind: PrometheusRule
metadata:
  creationTimestamp: "2024-04-29T23:53:40Z"
  generation: 1
  labels:
    app: netobserv-ebpf-agent
  name: ebpf-agent-prom-alert
  namespace: netobserv-privileged
  ownerReferences:
  - apiVersion: flows.netobserv.io/v1beta2
    blockOwnerDeletion: true
    controller: true
    kind: FlowCollector
    name: cluster
    uid: 7376e577-a0d4-4ffd-a83d-cacc0199e14e
  resourceVersion: "11150"
  uid: 51e833bc-aef1-4f9c-8210-4eea4500acd9
spec:
  groups:
  - name: NetobservEBPFAgentAlerts
    rules:
    - alert: NetObservNoFlows
      annotations:
        description: NetObserv eBPF agent hashmap table is full, it means that the
          eBPF agent is not able to process new flows. Please consider to increase
          the hashmap table size.
        summary: NetObserv eBPF is not able to process any new flows
      expr: sum(rate(netobserv_agent_dropped_flows_total[1m])) == 0
      for: 10m
      labels:
        app: netobserv
        severity: warning

```

File: controllers/ebpf/agent-metrics.go [hunk 1]
```diff
@@ -103,3 +113,54 @@ func (c *AgentController) promServiceMonitoring(target *flowslatest.FlowCollecto
...
+	// EBPF hashmap table is full Not receiving any new flows
+	if shouldAddAlert(flowslatest.AlertDroppedFlows, target.Metrics.DisableAlerts) {
+
+		rules = append(rules, monitoringv1.Rule{
+			Alert: string(flowslatest.AlertDroppedFlows),
+			Annotations: map[string]string{
+				"description": "NetObserv eBPF agent hashmap table is full, it means that the eBPF agent is not able to process new flows. Please consider to increase the hashmap table size.",
+				"summary":     "NetObserv eBPF is not able to process any new flows",
+			},
+			Expr: intstr.FromString("sum(rate(netobserv_agent_dropped_flows_total[1m])) == 0"),
```
Comment (jpinsonneau) [inline controllers/ebpf/agent-metrics.go:130]:
```suggestion
			Expr: intstr.FromString("sum(rate(netobserv_agent_dropped_flows_total[1m])) > 0"),
```

**Comment by msherif1234** [comment]
emulate hashmap table full alert
- set sampling to `1`
- reduce hashmap table size to `100`
- enable ebpf agent metric
- use `hey-ho` tool to generate traffic `./hey-ho.sh -r 5 -d 3 -z 10m -n 4 -q 2 -p -b`

image

**Review by jpinsonneau — APPROVED** [review]
Code looks good ! Thanks @msherif1234

**Comment by memodi** [comment]
@msherif1234 I deployed the operator bundle included in catalog [image](https://github.com/netobserv/network-observability-operator/pull/632#issuecomment-2084943950) , however I don't see the prometheus rule being configured in `netobserv-privileged` NS, am I missing something?
```
$ oc get prometheusrules -A  | egrep netobserv
netobserv                                    flowlogs-pipeline-alert                        78s
```

image

**Comment by memodi** [comment]
oh wait, I had ebpf metrics disabled, enabling it now.

**Pull request merged**